### DISCUSSION TITLE
Architecture/API reference?

### DISCUSSION THREAD

**Discussion by seangrogg** [mastodon/mastodon#33143] [Q&A]
Is there a reference architecture/API reference for those wanting to build out their own mastodon server implementation?

I have been incrementally working on my own mastodon implementation with an aim of producing something relatively lightweight for a single-user mostly-lurking external/minimal-client use case that would run well on a spare machine (and accepting that I may occasionally wink out of existence because I'm not a data center). A lot of progress I've made has largely been "black boxing" from using clients like `elk.zone` (huge thanks) and working backwards from failed status codes and combing through minified javascript for errors.

That said, I'd imagine this approach will start to show its limits when I start working on federation features (sending/receiving likes, comments, follows, etc). Most of the documentation I've read has largely looked at aspects of the server's implementation from the perspective of someone operating the server or developing a client (and sensibly so!). However, most of my search for API/architecture details have yielded relatively little so far. Am I missing some potentially valuable resources for implementing my own server? If not, are there any recommended resources or mastodon source files to study as I start to work on federation compatibility?

Huge thanks in advance!

**Comment by mjankowski** [comment]
You are not missing something. The API docs on the documentation site are probably the best thing, but they are not necessarily totally thorough and you may have to occasionally jump into the actual app code to fully see how something will work.
